pkg/controller/servicecidrs/servicecidrs_controller.go

```diff
@@ -0,0 +1,641 @@
+				// and is the only existing one, at least one IPAddress will be
+				// orphan, block the ServiceCIDR deletion.
+				ipPrefix, err := ipAddressToPrefix(ip)
+				if err != nil {
+					// the IPAddress object validates the name is a valid IPAddress
+					logger.Info("[SHOULD NOT HAPPEN] unexpected error parsing IPAddress", "IPAddress", ip.Name, "error", err)
+					continue
+				}
+				cidrs := sets.New[string]()
+				c.tree.WalkPath(ipPrefix, func(k netip.Prefix, v cidrSet) bool {
+					if prefixContainIP(k, ipPrefix.Addr()) {
+						cidrs.Insert(v.UnsortedList()...)
+					}
+					return false
+				})
+				if cidrs.Len() == 1 && cidrs.Has(cidr.Name) {
+					return false, nil
+				}
+				logger.V(4).Info("No IPv6 addresses referencing", "ServiceCIDR", cidr.String())
+			}
+		}
+	}
+	// There are no IPAddresses that depend on the existing ServiceCIDR, so
+	// it is safe to delete, remove finalizer.
+	logger.Info("ServiceCIDR no longer have orphan IPs", "ServiceCDIR", cidr.String())
+	return true, nil
+}
+
+func (c *Controller) addServiceCIDRFinalizerIfNeeded(ctx context.Context, cidr *networkingapiv1alpha1.ServiceCIDR) error {
+	for _, f := range cidr.GetFinalizers() {
+		if f == ServiceCIDRProtectionFinalizer {
+			return nil
+		}
+	}
+
+	patch := map[string]interface{}{
+		"metadata": map[string]interface{}{
+			"finalizers": []string{ServiceCIDRProtectionFinalizer},
+		},
+	}
+	patchBytes, err := json.Marshal(patch)
+	if err != nil {
+		return err
+	}
+	_, err = c.client.NetworkingV1alpha1().ServiceCIDRs().Patch(ctx, cidr.Name, types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	klog.FromContext(ctx).V(4).Info("Added protection finalizer to ServiceCIDR", "ServiceCIDR", cidr.Name)
+	return nil
+
+}
+
+func (c *Controller) removeServiceCIDRFinalizerIfNeeded(ctx context.Context, cidr *networkingapiv1alpha1.ServiceCIDR) error {
+	found := false
+	for _, f := range cidr.GetFinalizers() {
+		if f == ServiceCIDRProtectionFinalizer {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return nil
+	}
+	patch := map[string]interface{}{
+		"metadata": map[string]interface{}{
+			"$deleteFromPrimitiveList/finalizers": []string{ServiceCIDRProtectionFinalizer},
+		},
+	}
+	patchBytes, err := json.Marshal(patch)
+	if err != nil {
+		return err
+	}
+	_, err = c.client.NetworkingV1alpha1().ServiceCIDRs().Patch(ctx, cidr.Name, types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	klog.FromContext(ctx).V(4).Info("Removed protection finalizer to ServiceCIDs", "ServiceCIDR", cidr.Name)
+	return nil
+}
+
+func ipAddressToPrefix(ip *networkingapiv1alpha1.IPAddress) (netip.Prefix, error) {
+	labelFamily, ok := ip.Labels[networkingapiv1alpha1.LabelIPAddressFamily]
+	if !ok {
+		return netip.Prefix{}, fmt.Errorf("IPAddress without required label: %s", networkingapiv1alpha1.LabelIPAddressFamily)
```

Comment:
Given that checking the label introduces another "can't happen, but..." situation, why not just check `netutils.IsIPv6String()`?

Reply:
because sometimes things that can not happen , just happen and this time we know it 😄